pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,734 @@
+	MaxAmountExceeded,
+	/// Emits when the loan can not be borrowed because the loan is written off
+	WrittenOffRestriction,
+	/// Emits when maturity has passed and borrower tried to borrow more
+	MaturityDatePassed,
+}
+
+/// Error related to loan borrowing
+#[derive(Encode, Decode, TypeInfo, PalletError)]
+pub enum WrittenOffError {
+	/// Emits when a write off action tries to write off the more than the policy allows
+	LessThanPolicy,
+}
+
+/// Error related to loan closing
+#[derive(Encode, Decode, TypeInfo, PalletError)]
+pub enum CloseLoanError {
+	/// Emits when close a loan that is not fully repaid
+	NotFullyRepaid,
+}
+
+// Portfolio valuation information.
+// It will be updated on these scenarios:
+//   1. When we are calculating portfolio valuation for a pool.
+//   2. When there is borrow or repay or write off on a loan under this pool
+// So the portfolio valuation could be:
+//	 - Approximate when current time != last_updated
+//	 - Exact when current time == last_updated
+#[derive(Encode, Decode, Clone, Default, TypeInfo, MaxEncodedLen)]
+pub struct PortfolioValuation<Balance> {
+	// Computed portfolio valuation for the given pool
+	value: Balance,
+
+	// Last time when the portfolio valuation was calculated for the entire pool
+	last_updated: Moment,
+}
+
+impl<Balance> PortfolioValuation<Balance>
+where
+	Balance: tokens::Balance,
+{
+	pub fn new(value: Balance, when: Moment) -> Self {
+		Self {
+			value,
+			last_updated: when,
+		}
+	}
+
+	pub fn value(&self) -> Balance {
+		self.value
+	}
+
+	pub fn last_updated(&self) -> Moment {
+		self.last_updated
+	}
+
+	pub fn update_with_pv_diff(
+		&mut self,
+		old_pv: Balance,
+		new_pv: Balance,
+	) -> Result<(), ArithmeticError> {
+		match new_pv.cmp(&old_pv) {
+			Ordering::Greater => self.value.ensure_add_assign(new_pv.ensure_sub(old_pv)?),
+			Ordering::Less => self.value.ensure_sub_assign(old_pv.ensure_sub(new_pv)?),
+			Ordering::Equal => Ok(()),
+		}
+	}
+}
+
+/// Information about how the portfolio valuation was updated
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PortfolioValuationUpdateType {
+	/// Portfolio Valuation was fully recomputed to an exact value
+	Exact,
+	/// Portfolio Valuation was updated inexactly based on loan status changes
+	Inexact,
+}
+
+/// The data structure for storing a specific write off policy
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct WriteOffState<Rate> {
+	/// Number in days after the maturity has passed at which this write off policy is valid
+	pub overdue_days: u32,
+
+	/// Percentage of present value we are going to write off on a loan
+	pub percentage: Rate,
+
+	/// Additional interest that accrues on the written off loan as penalty
+	pub penalty: Rate,
+}
+
+impl<Rate> WriteOffState<Rate>
+where
+	Rate: FixedPointNumber,
+{
+	fn is_overdue(&self, maturity_date: Moment, now: Moment) -> Result<bool, ArithmeticError> {
```

Comment:
I would still rename this one!

Reply:
Then it should be `is_not_active` or something like that. Any naming idea?

Reply:
maybe `applicable`